src/gt4py/next/common.py

```diff
@@ -117,10 +129,102 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:
             else:
                 raise IndexError("UnitRange index out of range")
 
+    def __and__(self, other: UnitRange) -> UnitRange:
+        start = max(self.start, other.start)
+        stop = min(self.stop, other.stop)
+        return UnitRange(start, stop)
+
 
 DomainT: TypeAlias = tuple[tuple[Dimension, UnitRange], ...]
```

Comment:
Delete and change all uses of DomainT to the new domain, e.g. test_nd_array_field